src/hooks/useSelector.js

```diff
@@ -20,13 +20,24 @@ function useSelectorWithStoreAndSubscription(
   ])
 
   const latestSubscriptionCallbackError = useRef()
-  const latestSelector = useRef()
-  const latestStoreState = useRef()
-  const latestSelectedState = useRef()
 
   const storeState = store.getState()
+
+  const latestSelector = useRef(selector)
+  const latestStoreState = useRef(storeState)
+  const latestSelectedState = useRef()
+
   let selectedState
 
+  // lazy initialization pattern on first render
+  if (latestSelectedState.current === undefined) {
+    selectedState = selector(storeState)
+    latestSelectedState.current = {
+      first: true, // mark that this is the first render, so we can avoid calling selector in checkForUpdates
+      val: selectedState,
+    }
```

Comment:
Maybe weird that this is an object, I guess this could be split out into two refs?